modules/rpc/yarpc.go

```diff
@@ -36,67 +35,192 @@ import (
+	register  registerServiceFunc
+	config    yarpcConfig
+	log       ulog.Log
+	stateMu   sync.RWMutex
+	isRunning bool
 }
 
 var (
-	_dispatcherFn = defaultYarpcDispatcher
+	_dispatcherFn = defaultYARPCDispatcher
+	_dispatcherMu sync.Mutex
 
-	_ service.Module = &YarpcModule{}
+	_ service.Module = &YARPCModule{}
+
+	// A collection of all YARPC configs that are stored together to create
+	// a shared dispatcher. The YARPC team advised it to be a 'singleton'
+	// to control the lifecycle of all of the in/out bound traffic.
+	_configs configCollection
 )
 
-type registerServiceFunc func(module *YarpcModule)
+type registerServiceFunc func(module *YARPCModule)
 
 type yarpcConfig struct {
 	modules.ModuleConfig
-	Bind          string `yaml:"bind"`
-	AdvertiseName string `yaml:"advertiseName"`
+	Bind                    string
+	AdvertiseName           string `yaml:"advertiseName"`
+	inboundMiddleware       []middleware.UnaryInbound
+	onewayInboundMiddleware []middleware.OnewayInbound
+	inbounds                []transport.Inbound
+}
+
+// Stores a collection of all modules configs with a shared dispatcher
+// that are safe to call from multiple go routines. All the configs must
+// share the same AdvertiseName and represent a single service.
+type configCollection struct {
+	// sync configs
+	sync.RWMutex
+
+	// sync start/stop errors
+	errorMu sync.RWMutex
+
+	// idempotent start and stop
+	start      sync.Once
+	stop       sync.Once
+	stopError  error
+	startError error
+
+	configs    []*yarpcConfig
+	dispatcher *yarpc.Dispatcher
+}
+
+// Adds the config to the collection, the first config sets up the AdvertiseName requirement on consequent additions.
+func (c *configCollection) addConfig(config yarpcConfig) error {
+	c.Lock()
+	defer c.Unlock()
+
+	if len(c.configs) > 0 && config.AdvertiseName != c.configs[0].AdvertiseName {
+		return errs.Errorf("name mismatch, expected: %s; actual: %s",
+			c.configs[0].AdvertiseName,
+			config.AdvertiseName,
+		)
+	}
+
+	c.configs = append(c.configs, &config)
+	return nil
+}
+
+// Adds the default middleware: fx.context propagation and auth.
+func (c *configCollection) addDefaultMiddleware(host service.Host) error {
+	cfg := yarpcConfig{
+		AdvertiseName: host.Name(),
+		inboundMiddleware: []middleware.UnaryInbound{
+			fxContextInboundMiddleware{host},
+			authInboundMiddleware{host},
+		},
+		onewayInboundMiddleware: []middleware.OnewayInbound{
+			fxContextOnewayInboundMiddleware{host},
+			authOnewayInboundMiddleware{host},
+		},
+	}
+
+	if err := c.addConfig(cfg); err != nil {
+		host.Logger().Error("Can't add the default middleware to configs", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// Starts the dispatcher: wait until all modules call start, create a single dispatcher and then start it.
+// Once started the collection will not start the dispatcher again.
+func (c *configCollection) Start(host service.Host) error {
+	c.errorMu.Lock()
+	defer c.errorMu.Unlock()
```

Comment:
We do, otherwise we'll lose consistency:
A calls Start() and start executing Dispatcher start, but haven't finished before B calls Start().
Now B will return a nil error, while it can be not true.